### PR TITLE
feat(terminal)!: add inline viewport

### PR DESCRIPTION
**Pull request by conradludgate**
originally taken from https://github.com/fdehau/tui-rs/pull/552.

Updates `Viewport` type, making it an enum with `Fullscreen`, `Inline` and `Fixed` views.
`Inline(_)` includes the height of the inline screen.

Adds an `insert_before` method that can write permanent lines above the inline tui. Intended usecase is like `yarn install`, where there would be multiple updating elements at the bottom of the screen, and then 'xxx installed' would be permanent logs above the updating elements.

I've tried to minimise the breaking changes from the original PR, but there's still a breaking change in the `Backend` trait. In theory I can add some default methods to make it not a breaking change 

(CC @pdecat, https://github.com/ellie/atuin/pull/648)

### CODE REVIEW

**Comment by mindoodoo** [comment]
I just checked out the example, seems either bugged / unfinished or I'm just not understanding what @fdehau was trying to do. On my machine the example either wrote on top of itself on my occasions, or failed to "coexist" with previous outputs in my terminal. I'll try and take a look at this PR when I have the time, but I'll be pretty busy these upcoming days.

**Comment by conradludgate** [comment]
> seems either bugged / unfinished

my fault, push incoming

File: src/backend/mod.rs
Comment (conradludgate) [inline src/backend/mod.rs:40]: Assuming we don't want to break this trait, we could do
```diff
     fn draw<'a, I>(&mut self, content: I) -> Result<(), io::Error>
     where
         I: Iterator<Item = (u16, u16, &'a Cell)>;
-    fn append_lines(&mut self, n: u16) -> io::Result<()>;
+    fn append_lines(&mut self, _n: u16) -> io::Result<()> {
+        Ok(())
+    }
     fn hide_cursor(&mut self) -> Result<(), io::Error>;
     fn show_cursor(&mut self) -> Result<(), io::Error>;
     fn get_cursor(&mut self) -> Result<(u16, u16), io::Error>;
     fn set_cursor(&mut self, x: u16, y: u16) -> Result<(), io::Error>;
-    fn clear(&mut self, clear_type: ClearType) -> Result<(), io::Error>;
+    fn clear(&mut self) -> Result<(), io::Error>;
+    fn clear_to(&mut self, clear_type: ClearType) -> Result<(), io::Error> {
+        match clear_type {
+            ClearType::All => self.clear(),
+            ClearType::AfterCursor
+            | ClearType::BeforeCursor
+            | ClearType::CurrentLine
+            | ClearType::UntilNewLine => Err(io::Error::new(
+                io::ErrorKind::Other,
+                format!("clear_type [{clear_type:?}] not supported with this backend"),
+            )),
+        }
+    }
     fn size(&self) -> Result<Rect, io::Error>;
     fn flush(&mut self) -> Result<(), io::Error>;
 }
```

But I'm not sure if that would add confusion

Comment (mindoodoo) [reply src/backend/mod.rs:40]: Yeah I definitely don't think we want to break the trait.
But I don't think ```clear_to``` is the best name for the new function although to be fair I can't come up with anything better.
You could either leave it as is or if you find a better name change it. Maybe add a doc line that explains that ```clear``` will ClearType::All where as  ```clear_to``` can be specified.

Maybe ```clear_specified``` ?

Comment (ellie) [reply src/backend/mod.rs:40]: `clear_to` with documentation would work, though I also think `clear_area` or `clear_region` might also make sense?

Comment (gibbz00) [reply src/backend/mod.rs:40]: `clear_region` implemented in #119! :)

Comment (conradludgate) [reply src/backend/mod.rs:40]: clear_region works for me too! I'll update my PR tomorrow. I'll also add some documentation. Won't hurt

**Comment by ellie** [comment]
Hey! I don't want to pressure or anything, but the next [Atuin](https://github.com/ellie/atuin) release is blocked on this one getting in. Is there anything I can do to help

**Review by sayanarijit — APPROVED** [review]
LGTM. But this looks like a breaking change.

**Review by orhun — APPROVED** [review]

File: src/terminal.rs
Comment (sayanarijit) [inline src/terminal.rs:20]: Let's also implement pub fn inline(u16) and fullscreen() constructors...

Comment (sayanarijit) [reply src/terminal.rs:20]: @conradludgate

Comment (conradludgate) [reply src/terminal.rs:20]: Actually I'd rather delete this one. It's commented as "unstable" so it shouldn't be in use anyway, and it would be a small change of
```diff
- Viewport::fixed(n)
+ Viewport::Fixed(n)
```

Comment (sayanarijit) [reply src/terminal.rs:20]: Sounds good, as long as it's consistent.

**Review by orhun — APPROVED** [review]

**Review by mindoodoo — APPROVED** [review]

**Comment by orhun** [comment]
Thanks @hermannm! I'm currently having a chat with the other maintainers about the next possible release. We certainly don't want to keep people waiting 🐻

**Pull request merged**
